pkg/api/metrics.go

```diff
@@ -97,67 +97,63 @@ func (hs *HTTPServer) QueryMetricsV2(c *models.ReqContext, reqDTO dtos.MetricReq
 	return toMacronResponse(resp)
 }
 
-func toMacronResponse(qdr *backend.QueryDataResponse) response.Response {
-	statusCode := http.StatusOK
-	for _, res := range qdr.Responses {
-		if res.Error != nil {
-			statusCode = http.StatusBadRequest
-		}
-	}
-
-	return response.JSONStreaming(statusCode, qdr)
-}
+func (hs *HTTPServer) getDataSourceFromQuery(c *models.ReqContext, query *simplejson.Json, history map[string]*models.DataSource) (*models.DataSource, response.Response) {
+	var err error
+	uid := query.Get("datasource").Get("uid").MustString()
 
-// handleExpressions handles POST /api/ds/query when there is an expression.
-func (hs *HTTPServer) handleExpressions(c *models.ReqContext, reqDTO dtos.MetricRequest) response.Response {
-	timeRange := plugins.NewDataTimeRange(reqDTO.From, reqDTO.To)
-	request := plugins.DataQuery{
-		TimeRange: &timeRange,
-		Debug:     reqDTO.Debug,
-		User:      c.SignedInUser,
-		Queries:   make([]plugins.DataSubQuery, 0, len(reqDTO.Queries)),
+	// before 8.3 special types could be sent as datasource (expr)
+	if uid == "" {
+		uid = query.Get("datasource").MustString()
```

Comment:
Isn't this the data source name rather than uid?

Reply:
:thinking: true in dashboards, but for the metrics request it was weirdly set to `__expr__`

This endpoint was never sent name, rather it primarily worked with `datasourceId: int64`

Reply:
Understood. And now you'll send `__expr__` in both `uid` and `type` of `DataSourceRef`?

Reply:
yup -- *should* be just `type`, but right now enough things only look at uid that I don't want to muck with that just yet.  Ideally we can have defaults scoped to type rather than global.  ie "give me my default loki instance" -- but we are not there yet!